### PR TITLE
[AArch64] Implement spill/fill of predicate pair register classes

### PR DESCRIPTION
**Pull request by momchil-velikov**
We are getting ICE with, e.g.
```
#include <arm_sve.h>

 void g();
 svboolx2_t f0(int64_t i, int64_t n) {
     svboolx2_t r = svwhilelt_b16_x2(i, n);
     g();
     return r;
 }
```

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -2221,6 +2222,7 @@ unsigned AArch64InstrInfo::isStoreToStackSlot(const MachineInstr &MI,
   case AArch64::STRDui:
   case AArch64::STRQui:
   case AArch64::STR_PXI:
+  case AArch64::STR_PPXI:
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:2225]: Which test does this change affect?

Given that this store takes two P registers as operands, are the operand numbers below correct? (i.e. is MI.getOperand(1).isFI() ever true for STR_PPXI?

Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:2225]: Hmm, very likely they are here by mistake. I'll retest and will probably remove these two changes.

Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:2225]: Actually, I think the changes are correct and perhaps LDR/STR_ZZXI/etc need to be added here as well.
The store takes a single register as operand (representing a register pair).
Perhaps, I'll be able to construct a test case showing how it makes a difference, say in MachineLICM or in StackSlotColoring.

Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:2225]: Is it required though? or does it provide extra optimisation abilities? If it's the latter then probably best done as a follow patch.

Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Done.

File: llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td [hunk 2]
```diff
@@ -2378,11 +2378,13 @@ let Predicates = [HasSVEorSME] in {
     def LDR_ZZXI   : Pseudo<(outs   ZZ_b_strided_and_contiguous:$Zd), (ins GPR64sp:$sp, simm4s1:$offset),[]>, Sched<[]>;
     def LDR_ZZZXI  : Pseudo<(outs  ZZZ_b:$Zd), (ins GPR64sp:$sp, simm4s1:$offset),[]>, Sched<[]>;
     def LDR_ZZZZXI : Pseudo<(outs ZZZZ_b_strided_and_contiguous:$Zd), (ins GPR64sp:$sp, simm4s1:$offset),[]>, Sched<[]>;
+    def LDR_PPXI   : Pseudo<(outs PPR2:$pp), (ins GPR64sp:$sp, simm4s1:$offset),[]>, Sched<[]>;
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:2381]: `simm4s1` seems weird given the underlying instructions use `simm9` but then we don't target the pseudos via ISEL so I guess it doesn't really matter.  It's also consistent with the existing pseudos, which also look wrong.

Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:2381]: These offsets are checked when replacing frame index with actual addressing mode, so we know we are in range. I would suggest changing all these to `simm9` in a separate patch, if only to not make someone wonder why did we use such strange constraints.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 3]
```diff
@@ -3771,6 +3771,13 @@ bool AArch64InstrInfo::getMemOpInfo(unsigned Opcode, TypeSize &Scale,
     MinOffset = -256;
     MaxOffset = 255;
     break;
+  case AArch64::LDR_PPXI:
+  case AArch64::STR_PPXI:
+    Scale = TypeSize::getScalable(2);
+    Width = TypeSize::getScalable(2 * 2);
+    MinOffset = -256;
+    MaxOffset = 255;
```
[the remark this answers is not in the document]
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Done.

**Review by paulwalker-arm — APPROVED** [review]

**Pull request merged**